gnome3.gnome-settings-daemon: bypass polkit by using an suid wrapper

worldofpeace opened the pull request:
###### Motivation for this change
This is intended to fix #14168 untli we have a better solution from
upstream GNOME.

cc @jameysharp @hedning @jtojnar 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

worldofpeace committed:
nixos/gnome-settings-daemon: init

worldofpeace commented:
:thinking: Is pruning the polkit service away needed so it's not in the enviroment?

jtojnar commented:
> :thinking: Is pruning the polkit service away needed so it's not in the environment?

What do you mean?

worldofpeace commented:
> > thinking Is pruning the polkit service away needed so it's not in the environment?
> 
> What do you mean?

Is removing `$out/share/polkit-1` needed?

jtojnar commented on pkgs/desktops/gnome-3/core/gnome-settings-daemon/default.nix:
```diff
@@ -32,6 +32,10 @@ stdenv.mkDerivation rec {
   ];
 
+  preFixup = ''
+    gappsWrapperArgs+=(--set GSD_BACKLIGHT_HELPER "/run/wrappers/bin/gsd-backlight-helper")
```
Maybe we could parametrize the package for this and then add something like:

```nix
    nixpkgs.overlays = [
      (self: super: {
        gnome3 = gnome3.override (gself: gsuper: {
          gnome-settings-daemon = gnome-settings-daemon.override { inherit (config.security) wrapperDir; };
        });
      })
    ];
```

It should be no-op for default `wrapperDir`.

worldofpeace replied:
Sorry I don't follow.

jtojnar commented:
I do not think removing it is necessary – single unused field in some hashmap will not cause any problem.

worldofpeace committed:
gnome3.gnome-settings-daemon: bypass polkit by using an suid wrapper

This is intended to fix #14168 untli we have a better solution from
upstream GNOME.

The issue in brief is caused by the fact that gnome-settings-daemon is
non-restartable and if the policy definitions change in any way
it won't propagate to the running system because of that.

Also take note that we're abusing a debugging envar
`GSD_BACKLIGHT_HELPER` as it was never intended
for use outside testing.

worldofpeace commented:
Did the thing, also built and ran a vm.

This probably needs to be tested on a running machine though.

worldofpeace commented:
> Looks like a sensible pair of commits to me! I have not tested it yet though and I could well have missed important details.
> 
> I have this vague feeling that the entire preFixup hook should only happen when the NixOS module is what's using the gsd package, since otherwise there's no /run/wrappers copy of the helper. I'm not sure if Hydra would build and cache the wrapped version then though? Either way, I wouldn't block merging this PR over that question.

I'll try to test this later today. I've abused debugging features like this in the past and it generally goes well.

grahamc commented:
@GrahamcOfBorg test gnome3

grahamc commented on nixos/modules/services/desktops/gnome3/gnome-settings-daemon.nix:
```diff
@@ -0,0 +1,44 @@
+
+    security.wrappers.gsd-backlight-helper.source = "${pkgs.gnome3.gnome-settings-daemon}/libexec/gsd-backlight-helper";
+
+    nixpkgs.overlays = [
```
I think we probably don't want to be setting overlays in NixOS modules.

worldofpeace replied:
I was thinking the same thing, but @jtojnar thought it would be an improvement.
Not sure if hydra will cache it and I don't have an details to support against it.

worldofpeace commented:
@GrahamcOfBorg build nixosTests.gnome3

worldofpeace commented:
So I just realized that the code that allows us to do this in is 3.31.
I could just patch this feature into the latest release.

worldofpeace commented:
Is part of https://github.com/NixOS/nixpkgs/pull/57027
